02-intermediate-ruby/debugging.md

```diff
@@ -0,0 +1,108 @@
+# Debugging
+
+## Learning Goals
+- Read and analyze a stack trace
+- Understand how to use `pry` for debugging in Ruby code
+- Better understand the scope of variables and instances in an application when testing
+
+### Reading Errors via Stack Traces
+
+When an exception is raised, it immediately stops the current method, just like `return`. It will then bubble its way up through the program, method by method. If it makes it all the way to the top, Ruby will print out a summary of the exception and a description of what the program was doing when the exception happened, also known as a **stack trace**.
+
+A stack trace contains a ton of useful information, including a list of methods and blocks that Ruby was inside when the exception was raised. Being able to quickly read a complex stack trace is a super useful skill, so let's examine one now.
+
+First, take a look at the following stack trace that came from running some of the tests within the grocery store project.
+
+![test stack trace](images/test-stack-trace.png)
+
+**Activity**: Now, with your pair, spend a few minutes looking at this error in detail. We know you don't have access to the code, but try anyway to answer the following questions:
+1. What file has the code that was _originally called_ to create the error?
+2. What file has the code where the _error is located_? (Is this different than your answer to the question above?)
+3. Without seeing the code but using the error details, what do you think the code in question could look like?
+
+**Review**: While you don't have the code, an error message with a stack trace gives you a lot more information than you might be expecting!
+
+First, let's look at each piece of the provided error.
+
+![test stack trace errors](images/test-stack-trace-markup-p1.png)
+
+Next, we'll take a look at all of the details of the stack trace itself.
+
+![test stack trace details](images/test-stack-trace-markup-p2.png)
```

Comment:
The dark red in these images is not going to show up well against the black on the projector.

Reply:
Unfortunately I can just re-edit the images and I'd have to do the edits all over again, so I'm going to leave them for now.